src/VizzyUtils/QueryIngestion/VizzyPrepare.ts

```diff
@@ -0,0 +1,125 @@
+import {FormattedQuery, VisQueryResponse, VisData, QueryFormatterRules} from './types'
+
+export interface QueryFormatter {
+  /**
+   * Creates a slim, data-ful queryResponse. This function takes in the Looker query objects and returns a structure well suited for visualization development
+   * @param data - data obj from Looker
+   * @param config - config obj from Looker
+   * @param queryResponse - queryResponse obj from Looker
+   * @param vis - the vis obj itself
+   * @param rules - some options for error handling/boundary checking
+   */
+  prepare: (data: VisData, config: VisConfig, queryResponse: VisQueryResponse, vis: VisualizationDefinition, rules: QueryFormatterRules) => FormattedQuery,
+  /**
+   * Retrieves the requested field metadata
+   * @param queryFields - which queryResponse fields to consider
+   * @param columns - first N fields to get
+   */
+  getQueryMetadata: (queryFields: any, columns: number) => any,
+  /**
+   * Ensure that the prepared data shape is as expected
+   * @param data - single category data, from getQueryMetadata()
+   * @param addError - function for reporting an error back to Looker
+   * @param queryLength - actual length of category data, from Looker queryResponse
+   * @param strictLength - whether or not to enforce length check
+   */
+  validateInputShape: (data: any, addError: any, queryLength: number, strictLength: boolean) => void,
+  /**
+   * Ensure that the input data is of a shape which can be properly prepared
+   * @param data - data obj from Looker
+   * @param queryResponse - queryResponse obj from Looker
+   * @param rules - some options for error handling/boundary checking
+   * @param addError - function for reporting an error back to Looker
+   */
+  validatePrepareShape: (data: any, queryResponse: VisQueryResponse, rules: QueryFormatterRules, addError: any) => void,
+}
+
+export const VizzyQueryFormatter: QueryFormatter = {
+  validateInputShape(data, addError, queryLength, strictLength) {
+    if (strictLength) {
+      if (data.length !== queryLength) {
+        addError({title: "Invalid input", message: `Expected ${data.length} ${data[0][0].field_category}s, found ${queryLength}.`})
+      }
+    }
+  },
+  validatePrepareShape(data, queryResponse, rules, addError) {
+    const actualDims = queryResponse.fields.dimension_like.length
+    const actualMeas = queryResponse.fields.measure_like.length
+    if (data.length === 0) {
+      addError({title: "No data"})
+    }
+    if (rules.strictDims && actualDims !== rules.numDims) {
+      addError({title: "Invalid input", message: `Expected ${rules.numDims} dimensions, found ${actualDims}.`})
+    }
+    else if (rules.strictDims && actualMeas !== rules.numMeas) {
+      addError({title: "Invalid input", message: `Expected ${rules.numMeas} measures, found ${actualMeas}.`})
+    }
+  },
+  getQueryMetadata(queryFields, columns) {
+    const filtered = queryFields.filter((d: any, i: number) => {
```

Comment:
more `any`, I won't comment on them from here on down, but suggest a thorough audit or just not using typescript if you plan to not leverage it.